Kernel/Devices/USB/UHCIController.cpp

```
@@ -312,62 +315,191 @@ void UHCIController::start()
+        if (packet_size > max_size) {
+            packet_size = max_size;
+        }
+
+        current_td = create_transfer_descriptor(pipe, direction, packet_size);
+        if (current_td == nullptr)
+            return false;
+
+        current_td->set_buffer_address(buff_addr + byte_count);
+        byte_count += packet_size;
+
+        if (prev_td != nullptr)
+            prev_td->insert_next_transfer_descriptor(current_td);
+        else
+            first_td = current_td;
+
+        prev_td = current_td;
+    }
+
+    *last_td = current_td;
+    *td_chain = first_td;
+    return true;
+}
+
+void UHCIController::free_descriptor_chain(TransferDescriptor* first_descriptor)
+{
+    TransferDescriptor* descriptor = first_descriptor;
+
+    while (descriptor) {
+        descriptor->free();
+        descriptor = descriptor->next_td();
+    }
+}
+
+size_t UHCIController::submit_control_transfer(USBTransfer& transfer)
+{
+    USBPipe& pipe = const_cast<USBPipe&>(transfer.pipe()); // Short circuit the pipe related to this transfer
+    bool direction_in = (transfer.request().request_type & USB_DEVICE_REQUEST_DEVICE_TO_HOST) == USB_DEVICE_REQUEST_DEVICE_TO_HOST;
+
+    // Create the setup transfer descriptor
+    TransferDescriptor* setup_td = create_transfer_descriptor(pipe, PacketID::SETUP, sizeof(USBRequestData));
+    setup_td->set_buffer_address(reinterpret_cast<intptr_t>(transfer.buffer_physical().as_ptr()));
+
+    // Create a new descriptor chain
+    TransferDescriptor* last_data_descriptor;
+    TransferDescriptor* data_descriptor_chain;
+    bool did_create_chain = create_chain(pipe,
+        direction_in ? PacketID::IN : PacketID::OUT,
+        reinterpret_cast<intptr_t>(transfer.buffer_physical().as_ptr() + sizeof(USBRequestData)),
+        pipe.max_packet_size(),
+        transfer.transfer_data_size(),
+        &data_descriptor_chain,
+        &last_data_descriptor);
+
+    if (!did_create_chain)
+        return 0; // Transfer failed
```

This is really opaque. Let's make `create_chain` return `KResult` so we can propagate any errors properly.